### DISCUSSION TITLE
CUDA enabled version installation problem

### DISCUSSION THREAD

**Discussion by Ryan12345678910111213** [IBM/aihwkit#522] [Installation Q&A]
[error 5.txt](https://github.com/IBM/aihwkit/files/12033056/error.5.txt)
When I tried to do CMAKE_PREFIX_PATH=$CONDA_PREFIX make build_cuda flags="-DRPU_CUDA_ARCHITECTURES='86'", I got the error message in the error5.txt

Pytorch version: 2.0.1 torchvision 0.15.2
Package version: aihwkit 0.7.1
OS: Ubuntu 22.04.2 LTS (GNU/Linux 5.15.90.1-microsoft-standard-WSL2 x86_64)
Python version: 3.8.17
NVIDIA GeForce RTX 3060 Laptop
Cuda compilation tools, release 11.5, V11.5.119

**Comment by maljoras** [comment]
It looks like that it cannot find a properly installed torch. Can you try to open python and try 'import torch'?

Or better still try: 
```
/home/yuz422/venv/bin/python  -c "import torch; print('1' if torch._C._GLIBCXX_USE_CXX11_ABI else '0')"
```
If that does not work, you might have two python installations and installed the torch under the wrong one.

**Comment by Ryan12345678910111213** [comment]
@maljoras 
(venv) (base) yuz422@LAPTOP-U0NG45HI:~/aihwkit$ python
Python 3.8.17 (default, Jun  6 2023, 20:10:50)
[GCC 11.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ModuleNotFoundError: No module named 'torch'
>>>

It does not work. This is strange because I had already installed pytorch using conda install -c pytorch in miniconda. Could you please give me more suggestions?

**Comment by maljoras** [comment]
You have `venv` and `base` activated. Not sure what you are doing there. Typically you have either venv or miniconda installed. You are apparently activating the venv where your python is stored but if you do `conda install` then it will store it under conda. You could use `python -m pip install torch` then it would install it under your current environment.   But it is not a good idea to have two pythons installed at the same time.

**Reply by Ryan12345678910111213** [reply]
@maljoras 
[error6.txt](https://github.com/IBM/aihwkit/files/12034292/error6.txt)

Thank you very much for your help. 

```python
`(venv) (base) yuz422@LAPTOP-U0NG45HI:~/aihwkit$ python --version
Python 3.8.17
(venv) (base) yuz422@LAPTOP-U0NG45HI:~/aihwkit$ python3 --version
Python 3.8.17
(venv) (base) yuz422@LAPTOP-U0NG45HI:~/aihwkit$ conda info --envs
# conda environments:
#
base                  *  /home/yuz422/anaconda3
aihwkit                  /home/yuz422/anaconda3/envs/aihwkit`
``` 

I checked my python, but I didn't find that I had two Pythons installed. Could you please tell me how did you find this?
After I used python -m pip install torch to install the torch, I tried to do CMAKE_PREFIX_PATH=$CONDA_PREFIX make build_cuda flags="-DRPU_CUDA_ARCHITECTURES='86'" again. But  I got the same error message like before.

**Reply by maljoras** [reply]
That looks like a GCC / CUDA compiler issue special to CUDA toolkit version 11.5. See [here](https://github.com/NVlabs/instant-ngp/issues/119#issuecomment-1113979878) 
 
It was solved by updating the cuda toolkit to a newer version than 11.5. For instance, I am using CUDA 12.2 with GCC 11.3 and it works fine. But any version other than 11.5 are working, too. 11.7 is the current version which torch uses, I think.

**Reply by Ryan12345678910111213** [reply]
@maljoras 
Ok, I will try the way you mentioned.
Besides, I tried to use the instruction $conda install python=3.9 aihwkit-gpus -c conda-forge -c  file:/$HOME/aihwkit-condapkg to install conda packages on the department computer, I encountered an error:
```python
(aihwkit) yuz422@ece-hpc06:~ $ conda install python=3.9 aihwkit-gpu -c conda-forge -c file:/home/yuz422/aihwkit-condapkg/
Solving environment: failed

CondaHTTPError: HTTP 404 NOT FOUND for url <https://conda.anaconda.org/file:/home/yuz422/aihwkit-condapkg/noarch/repodata.json>
Elapsed: 00:00.085190
CF-RAY: 7e66549a2ab20cc0-EWR

The remote server could not find the noarch directory for the
requested channel with url: https://conda.anaconda.org/file:/home/yuz422/aihwkit-condapkg

As of conda 4.3, a valid channel must contain a `noarch/repodata.json` and
associated `noarch/repodata.json.bz2` file, even if `noarch/repodata.json` is
empty. please request that the channel administrator create
`noarch/repodata.json` and associated `noarch/repodata.json.bz2` files.
$ mkdir noarch
$ echo '{}' > noarch/repodata.json
$ bzip2 -k noarch/repodata.json

You will need to adjust your conda configuration to proceed.
Use `conda config --show channels` to view your configuration's current state.
Further configuration help can be found at <https://conda.io/docs/config.html>.


(aihwkit) yuz422@ece-hpc06:~ $

``` 
Could you please give me some suggestions to solve this?

**Reply by maljoras** [reply]
For the conda package, you need to download the file first. As you can clearly see from the error message it cannot resolve the file path as it is interpreted as an malformed URL.

**Reply by Ryan12345678910111213** [reply]
Yeah, that's what I find strange. In fact, I've already downloaded the conda package and extracted it to the path /home/yuz422/aihwkit-condapkg.

**Reply by maljoras** [reply]
Could be that you need `file:///home/yuz422/aihwkit-condapkg/` instead of `file:/home/yuz422/aihwkit-condapkg/`

**Reply by maljoras** [reply]
Indeed, it needs to have `file:///` (3x `/`). It is wrong in the documentation, we will fix it.

**Reply by Ryan12345678910111213** [reply]
(1) I have completed the whole installation process on my own computer, but I still can't run the program. I can't find this error on the Internet. Could you please give me some advice?
```python
(aihwkit_1) (aihwkit_envgpu) yuz422@LAPTOP-U0NG45HI:~/code$ python3 test.py
Traceback (most recent call last):
  File "/home/yuz422/code/test.py", line 4, in <module>
    from aihwkit.nn import AnalogLinear
  File "/home/yuz422/anaconda3/envs/aihwkit_1/lib/python3.9/site-packages/aihwkit/nn/__init__.py", line 17, in <module>
    from aihwkit.nn.modules.container import AnalogSequential
  File "/home/yuz422/anaconda3/envs/aihwkit_1/lib/python3.9/site-packages/aihwkit/nn/modules/container.py", line 24, in <module>
    from aihwkit.nn.modules.base import AnalogModuleBase
  File "/home/yuz422/anaconda3/envs/aihwkit_1/lib/python3.9/site-packages/aihwkit/nn/modules/base.py", line 26, in <module>
    from aihwkit.simulator.configs.utils import MappingParameter
  File "/home/yuz422/anaconda3/envs/aihwkit_1/lib/python3.9/site-packages/aihwkit/simulator/configs/__init__.py", line 15, in <module>
    from .configs import (
  File "/home/yuz422/anaconda3/envs/aihwkit_1/lib/python3.9/site-packages/aihwkit/simulator/configs/configs.py", line 18, in <module>
    from aihwkit.simulator.configs.devices import (
  File "/home/yuz422/anaconda3/envs/aihwkit_1/lib/python3.9/site-packages/aihwkit/simulator/configs/devices.py", line 21, in <module>
    from aihwkit.simulator.configs.helpers import (
  File "/home/yuz422/anaconda3/envs/aihwkit_1/lib/python3.9/site-packages/aihwkit/simulator/configs/helpers.py", line 20, in <module>
    from aihwkit.simulator.rpu_base import devices, tiles
ImportError: /home/yuz422/anaconda3/envs/aihwkit_1/lib/python3.9/site-packages/aihwkit/simulator/rpu_base.cpython-39-x86_64-linux-gnu.so: undefined symbol: _ZN3c106detail14torchCheckFailEPKcS2_jRKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE
``` 
(2) In the university department of computer use file:///home/yuz422/aihwkit-condapkg/ also still can't solve the problem.
```python
(aihwkit_1) yuz422@ece-hpc06:~ $ conda install python=3.9 aihwkit-gpu -c conda-forge -c file:///home/yuz422/aihwkit-condapkg
Solving environment: failed

PackagesNotFoundError: The following packages are not available from current channels:

  - aihwkit-gpu

Current channels:

  - https://conda.anaconda.org/conda-forge/linux-64
  - https://conda.anaconda.org/conda-forge/noarch
  - file:///home/yuz422/aihwkit-condapkg/linux-64
  - file:///home/yuz422/aihwkit-condapkg/noarch
  - https://repo.anaconda.com/pkgs/main/linux-64
  - https://repo.anaconda.com/pkgs/main/noarch
  - https://repo.anaconda.com/pkgs/free/linux-64
  - https://repo.anaconda.com/pkgs/free/noarch
  - https://repo.anaconda.com/pkgs/r/linux-64
  - https://repo.anaconda.com/pkgs/r/noarch
  - https://repo.anaconda.com/pkgs/pro/linux-64
  - https://repo.anaconda.com/pkgs/pro/noarch

To search for alternate channels that may provide the conda package you're
looking for, navigate to

    https://anaconda.org

and use the search bar at the top of the page.


(aihwkit_1) yuz422@ece-hpc06:~ $

```
